text/0000-project-descriptor-schema.md

```diff
@@ -0,0 +1,200 @@
+and validators in a wide variety of languages. Tools validating a project
+descriptor will need to convert the descriptor to JSON, then use schema to
+validate the JSON descriptor.
+
+The schema might look something like this:
+
+```json
+{
+  "$schema": "http://json-schema.org/schema#",
+  "$id": "https://buildpacks.io/schemas/project-descriptor.json",
+  "type": "object",
+  "properties": {
+    "build" : {
+      "type": "object",
+      "properties": {
+        "env": {
+          "type": "array",
+          "items": {
+            "type": "object",
+            "required": ["name", "value"],
+            "properties": {
+              "name": {
+                "type": "string"
+              },
+              "value": {
+                "type": "string"
+              }
+            }
+          }
+        },
+        "include": {
+          "type": "array",
+          "items": {
+            "type": "string"
+          }
+        },
+        "exclude": {
+          "type": "array",
+          "items": {
+            "type": "string"
+          }
+        },
+        "buildpacks": {
+          "type": "array",
+          "items": {
+            "type": "object",
+            "required": ["id"],
+            "properties": {
+              "id": { 
+                "type": "string"
+              },
+              "version": {
+                "type": "string"
+              },
+              "uri": { 
+                "type": "string"
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+}
+```
+
+This schema could be further enhanced to define the exclusivity of
+`build.include` and `build.exclude` and other complex rules.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- If the schema is published in an independent repository, it complicates the
+  dependency graph and changes to the schema become more complicated to release.
+- It may potentially break builds for users that have undected errors in their
+  function.toml.
+- JSON-Schema error messages are more informative than errors we have now, but 
+  can sometimes be cryptic as the schema complexity grows.
+
+# Alternatives
+[alternatives]: #alternatives
+
+- Do nothing
+  With this alternative, project descriptor errors continue to be a bit weird 
+  around the edges. Additionally, developing integrations with and extensions
+  to the project descriptor requires reimplementing parts of pack.
+- Further iterate on descriptor parsing in `pack`
+  This would involve adding validation rules to `pack` and improve error
+  messages. This is tougher for intergrations to work with, and doesn't scale
+  as well as more project descriptor features and rules are added.
+
+# Prior Art
+[prior-art]: #prior-art
+
+- [The project descriptor RFC](./0019-project-descriptor.md)
+
+# Unresolved Questions
+[unresolved-questions]: #unresolved-questions
+
+- What parts of the design do you expect to be resolved before this gets merged?
+  - Where should this schema live?
```

Comment:
I'd propose `github.com/buildpacks/project-toml`. But if we want to aim high, we could create a new Github org, and do something like `github.com/project-toml/schema`

Reply:
also `github.com/buildpacks/schemas` if we want it to include schemas for all our file formats

Reply:
What if, for all specified files, we put schemas, and go bindings directly in the `buildpacks/spec`?

Prior art: 
https://github.com/opencontainers/image-spec contains `schema` and `specs-go` directories

Reply:
This wouldn't be an appropriate place for unspecified schemas like `builder.toml`, but it seems appropriate for schemas describing `pack` specific files to live in the `pack` repo